### PR TITLE
Windows: fix path normalization in coqdep

### LINKED ISSUE #19579: [coqdep] path relativization doesn't work for some Windows setups
### Description of the problem

The root of this issue is the following dune bug: https://github.com/ocaml/dune/issues/5833

Due to this bug, Dune fails when a rule depends on an absolute path that is inside the workspace. This is a terrible bug, but indeed signals the difficulty of mapping absolute to relative paths and viceversa in Unix envs.

The timeline for this bug is,
- #15220: introduced findlib loading mode, but without coordination with Dune, several important and extremely time-consuming fixes came in #15991 , #16073, #16075, #16115, #16428, #16658
- After these fixes, and adding support for this mode to dune upstream (https://github.com/ocaml/dune/pull/6167), we were hit by the dune bug. The problem is that coqdep will query findlib for the path of the `.cmxs` file, however `findlib`  will always return an absolute path, triggering the bug.
- Rodolphe added a workaround in coqdep (#17270, #18165), that indeed will try to turn the absolute paths from `findlib` in relative paths.

It seems that two problems do remain:
- path relativization doesn't work properly on windows, due to `C:\` path format
- path relativization may be used on libraries that are not in the build workspace.

Gaëtan pushed a workaround for the second issue at https://github.com/coq/coq/pull/18385 , however I'm not sure that is enough.

IMHO, these issues are easy to fix properly (and add tests).

### PR DESCRIPTION
**Pull request by MSoegtropIMC**
Fixes / closes #19579

This is a simple fix I had identically in Coq 8.18. It didn't get merged into 8.19 because there was a rework in these files (the file functions were moved to a separate file file_utils) overlapping with my fixes and then it somehow got lost.

See https://github.com/coq/platform/blob/main/opam/opam-repository/packages/coq-core/coq-core.8.18.0/files/0001-ocamldep-fix-to_relative_path-on-Windows.patch

The identical fix is well tested in Coq Platform 8.18 and it also fixed the issue with coq-quickchick in Coq Platform for 8.19.2.

Also note that the same method is used in another function in the same file.

### CODE REVIEW

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by ejgallego** [comment]
cc: @rlepigre

**Comment by proux01** [comment]
@Lysxia itree seems broken on master

**Review by rlepigre — APPROVED** [review]
This looks good to me.

**Comment by proux01** [comment]
CI faliure is unrelated, let's merge
@coqbot merge now

**Comment by ejgallego** [comment]
I actually wonder if we should make this function return a `Result.t`, in the case that the paths are not relativizable.

I've seen this problem quite a few times around, in particular, I think that if we don't want to worry about this we got to add tests for the matrix of:

- unix style paths, cygwin (hybrid) style paths, windows style paths (both with network root and drive letter root)
- two files in the same build root, one file global one in the build workspace, both out

These are a non-trivial amount of tests, however IMHO that is important as the tests will save a lot of time in the future.

p.s: I've added some missing tags, including the need to test this.

**Comment by MSoegtropIMC** [comment]
@ejgallego : I don't think we actually need to support absolute cygwin paths (where / is something like C:\cygwin\) because afaik OCaml is still broken on cygwin so there doesn't exist a cygwin Coq. Coq Platform uses cygwin as build host, but the resulting executables are MinGW.

What are the cases of not relativizable paths? I guess this includes Paths to different Windows disks. Are there also cases under Linux or MacOS?

**Comment by MSoegtropIMC** [comment]
What is the official process to request a back merge to 8.20?

**Comment by proux01** [comment]
Before merge: use 8.20.1 as milestone
Now that it's merged, please open a PR targetting branch 8.20 (and ping release managers if they miss it, here @silene and me)

**Comment by MSoegtropIMC** [comment]
@proux01 : thanks, I will cherry pick it to the 8.20 branch and do a PR.

**Comment by ejgallego** [comment]
> @ejgallego : I don't think we actually need to support absolute cygwin paths (where / is something like C:\cygwin) because afaik OCaml is still broken on cygwin so there doesn't exist a cygwin Coq. Coq Platform uses cygwin as build host, but the resulting executables are MinGW.

@MSoegtropIMC , I am no expert, but as of today I ship two builds for coq-lsp + Windows (both of them fully functional and native): one with opam 2.2 and the other with the Coq Platform. In both cases, the paths that VSCode send to coq-lsp are of the form `C:\`.. etc... So I'm unsure what this means in this context, but keep in mind that the opam 2.2 dune build is very very native, it doesn't ever call bash for example (to build Coq, still opam has its copy for compat)

IMHO adding some tests would save from a lot of headaches in the future.

> What are the cases of not relativizable paths? I guess this includes Paths to different Windows disks. Are there also cases under Linux or MacOS?

That's one case indeed; in general, we should not relativise the paths that are outside of the build dir.

So if we have as build dir `/home/egallego/coq` and a path is `/home/egallego/.opam/foo`, the path is not relative to the build dir and should not be relativized.

I am not sure tho we have a way to determine the build dir tho. The dune bug is a PITA.

**Comment by ejgallego** [comment]
I would fix the dune bug as it would likely take me less than a day, but unfortunately I'm 100% booked until the end of my contract end of month :/

**Comment by MSoegtropIMC** [comment]
> @MSoegtropIMC , I am no expert, but as of today I ship two builds for coq-lsp + Windows (both of them fully functional and native): one with opam 2.2 and the other with the Coq Platform. In both cases, the paths that VSCode send to coq-lsp are of the form `C:\`.. etc... So I'm unsure what this means in this context

This are Windows paths. Cygwin paths are Linux style paths (like /home/Michael/coq) which are mapped by cygwins wrapper layer to a Windows path under the cygwin root.

The Coq Platform build is a cross build - it uses cygwin as build host and MinGW (native Windows) as target. It is similar to building a Windows app on Linux except that it is easier to run the created executables (like coqc). All executables created by and installed by Coq Platform are pure native Windows apps and accept and produce only Windows paths (but using / instead of \ is fine, which is true for all Windows programs - only the parser of CMD doesn't like this).

During Coq Platform build, shell scripts like configure scripts, make and m4 (autotools) are all cygwin and can only work with cygwin paths, which are compatible with WIndows paths using / as long as they are relative. The cygwin tools also can pass Windows paths say from `coqc - where` to a MinGW app, but e.g. make does not understand such paths.

It is a bit astonishing that this produces as little issues as it does, but in part this is due to the fact that we work since years on removing all incompatibilities. Still it was not that much effort to do this.

**Comment by MSoegtropIMC** [comment]
> I would fix the dune bug as it would likely take me less than a day, but unfortunately I'm 100% booked until the end of my contract end of month :/

There is no urgent need. Coq Platform builds are fine and this is what I am mostly interested in. It is also interesting to note that quickchick was the only package in extended Coq Platform suffering from this.

**Comment by ejgallego** [comment]
Indeed I am not sure how my setup is, I guess due to use vs code native for win, but all the paths we handle are classical win paths.

Even if the platform build, beware of the case where for example you have a plugin installed in opal, and you are building one other locally.

**Pull request merged**
